ovn: fix a deadlock in the NetworkPolicy code

danwinship opened the pull request:
If you add a NetworkPolicy containing a combined namespaceSelector and podSelector where there is already at least one pod that matches the selectors, then the thread will deadlock (trying to re-grab the `np` mutex it's already holding) while also holding the mutex on the `Pod` informer, so the master will stop processing both NetworkPolicy and Pod events.

This fixes the deadlock, and I think it's correct? It's small but it's not trivial...

Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1752222

@dcbw @girishmg @JacobTanenbaum

dcbw commented:
@danwinship like a normal Informer the WatchFactory will synchronously call Add() for each item in the list when the handler is added.

Another thought would be to change the WatchFactory code to *not* do this when a `processExisting` is passed (since that's basically redundant with the initial Add calls), then update `handlePeerNamespaceAndPodSelector()` to run the process existing from a gofunc?

danwinship commented:
@dcbw hm... I think that's a lot more complicated (and might end up having side effects elsewhere?) Are you ok with this approach?

I could also change the

				np.Lock()
				if np.deleted {
					np.Unlock()
					return
				}
				np.Unlock()

to:

				np.Lock()
				alreadyDeleted := np.deleted
				np.Unlock()
				if alreadyDeleted {
					return
				}

if that would be clearer...

dcbw commented:
@danwinship yeah slightly clearer I think.  In any case this approach is the best solution for now.

danwinship commented:
@dcbw updated

dcbw commented:
/lgtm
